File: aa-integration-tests/tests/fixtures/agents/python/agent_team/langgraph_team.py
"""Two-agent LangGraph team fixture — F116 E2E acceptance (AAASM-1513).

Scenario: agent_team / framework: langgraph
Registers a root agent + one team member in the same process.
"""

from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from _shared import emit, load_config


def run_real(cfg: dict) -> None:
    import asyncio
    from typing import TypedDict

    from langgraph.graph import END, StateGraph

    from agent_assembly import init_assembly

    root_id = cfg["agent_id"] + "-root"
    member_id = cfg["agent_id"] + "-member"

    class State(TypedDict):
        task: str
        result: str

    def process_node(state: State) -> dict[str, str]:
        return {"task": state["task"], "result": f"processed: {state['task']}"}

    # init_assembly() supports one active context per process; each agent context is
    # shut down before the next is created so the global context slot is free.
    for agent_id, role in [(root_id, "root"), (member_id, "member")]:
        ctx = init_assembly(
            gateway_url=cfg["gateway_addr"],
            api_key="e2e-test-key",
            agent_id=agent_id,
            team_id="f116-e2e",
            mode="sdk-only",
        )
        asyncio.run(ctx.client.register_agent())
        emit({"event": "started", "agent_id": agent_id, "role": role, "framework": "langgraph"})
        # StateGraph.compile() is patched by LangGraphPatch; node wrappers fire governance hooks.
        graph = StateGraph(State)
        graph.add_node("process", process_node)
        graph.set_entry_point("process")
        graph.add_edge("process", END)
        graph.compile().invoke({"task": cfg["task"], "result": ""})
        ctx.shutdown()

    emit({"event": "tool_call", "tool": "langgraph_team_tool", "input": cfg["task"]})
    emit({"event": "done", "result": "langgraph agent_team", "agent_count": 2})


if __name__ == "__main__":
    cfg = load_config()
    if os.environ.get("AA_SELFTEST") == "1":
        root_id = cfg["agent_id"] + "-root"
        member_id = cfg["agent_id"] + "-member"
        emit({"event": "started", "agent_id": root_id, "role": "root", "framework": "langgraph"})
        emit({"event": "started", "agent_id": member_id, "role": "member", "framework": "langgraph"})
        emit({"event": "tool_call", "tool": "langgraph_team_tool", "input": ""})
        emit({"event": "done", "result": "selftest-ok", "agent_count": 2})
        sys.exit(0)
    run_real(cfg)

File: aa-integration-tests/tests/fixtures/agents/typescript/agent_team/langgraph_team.ts
// F116 ST-B — two-agent LangGraph team fixture (AAASM-1514).
//
// Registers a root agent, then builds a LangGraph where the coordinator node
// spins up a child assembly context (simulating agent delegation). In selftest
// mode emits synthetic events for hermetic CI runs.
//
// Invocation:
//   AA_GATEWAY_ADDR=127.0.0.1:PORT AA_AGENT_ID=e2e-lg-root \
//     pnpm exec tsx agent_team/langgraph_team.ts
//
//   AA_SELFTEST=1 AA_GATEWAY_ADDR=dummy pnpm exec tsx agent_team/langgraph_team.ts

import { loadConfig, emit, type AgentConfig } from "../_shared.js";
import { initAssembly } from "@agent-assembly/sdk";
import { StateGraph, Annotation } from "@langchain/langgraph";

const StateAnnotation = Annotation.Root({
  task: Annotation<string>(),
  results: Annotation<string[]>({ reducer: (a, b) => [...a, ...b], default: () => [] }),
});

async function runReal(cfg: AgentConfig): Promise<void> {
  const rootCtx = await initAssembly({
    gatewayUrl: `http://${cfg.gatewayAddr}`,
    apiKey: "e2e-test-key",
    agentId: cfg.agentId,
    teamId: "f116-e2e",
    mode: "sdk-only",
  });

  emit({ event: "started", agent_id: cfg.agentId, role: "coordinator" });

  const graph = new StateGraph(StateAnnotation)
    .addNode("coordinator", async (state) => {
      const workerCtx = await initAssembly({
        gatewayUrl: `http://${cfg.gatewayAddr}`,
        apiKey: "e2e-test-key",
        agentId: `${cfg.agentId}-worker`,
        teamId: "f116-e2e",
        parentAgentId: cfg.agentId,
        mode: "sdk-only",
      });
      emit({ event: "started", agent_id: `${cfg.agentId}-worker`, role: "worker" });
      await workerCtx.shutdown();
      return { results: [`worker: ${state.task}`] };
    })
    .addEdge("__start__", "coordinator")
    .addEdge("coordinator", "__end__")
    .compile();

  const output = await graph.invoke({ task: cfg.task });
  emit({ event: "tool_call", tool: "coordinator", input: cfg.task });

  await rootCtx.shutdown();
  emit({ event: "done", result: output.results.join(", ") });
}

const cfg = loadConfig();

if (process.env.AA_SELFTEST === "1") {
  emit({ event: "started", agent_id: cfg.agentId, role: "coordinator" });
  emit({ event: "started", agent_id: `${cfg.agentId}-worker`, role: "worker" });
  emit({ event: "tool_call", tool: "coordinator", input: cfg.task });
  emit({ event: "done", result: "selftest-ok" });
  process.exit(0);
}

await runReal(cfg);
